### ISSUE TITLE
Allow platform compatibility to be queried on packages

### ISSUE THREAD

**Issue by 3noch** [NixOS/nixpkgs#21845]
@cleverca22 pointed out:

```
nix-repl> (import <nixpkgs> { system = "x86_64-darwin"; }).acl.meta
error: Package ‘acl-2.2.52’ in ‘/nix/store/jk5dvrv6w9bcgh88g0x7clk19df3q28f-nixos-17.03pre96925.1c50bdd/nixos/pkgs/development/libraries/acl/default.nix:29’ is not supported on ‘x86_64-darwin’, refusing to evaluate.
```

This is unfortunate because it would be extremely handy to have a function that, given a package, could tell me if it is compatible with a system (most often, my current system). With such a function we could safely make certain features/dependencies optional without having to manually duplicate platform compatibility information with the likes of `isDarwin`, `isLinux`, etc.

**Comment by matthewbauer** [comment]
how about:

```
let
  isCompatible = pkg: builtins.elem builtins.currentSystem pkg.meta.platforms;
in
  isCompatible (import <nixpkgs>).acl
```

The problem I see in doing this in a derivation is that just because a dependency builds on a system doesn't mean that we always want it as an input in another derivation (see x11 libs in most GUI application on Darwin). These edge cases make things harder for a sort of automated builtInputs checker.

**Comment by RonnyPfannschmidt** [comment]
since nix is lazy to begin with, how about a way to express derivation that cannot materialize for a reason,

that way attributes may still be accessible, but one could request that reason and/or simply watch it fail to materialize

**Comment by vcunat** [comment]
That `isCompatible` function won't work, as evaluating `pkg.meta` throws the error already.  It would be best if we could avoid that, though it might be nontrivial to implement.

**Comment by matthewbauer** [comment]
I guess allowBroken would fix that?

**Comment by vcunat** [comment]
Fixed by 67e8392383.

**Comment by 3noch** [comment]
Yay!
